sql/core/src/main/scala/org/apache/spark/sql/execution/stat/StatFunctions.scala

```diff
@@ -77,4 +79,27 @@ private[sql] object StatFunctions {
       })
     counts.cov
   }
+
+  /** Generate a table of frequencies for the elements of two columns. */
+  private[sql] def crossTabulate(df: DataFrame, col1: String, col2: String): DataFrame = {
+    val tableName = s"${col1}_$col2"
+    val counts = df.groupBy(col1, col2).agg(col(col1), col(col2), count("*")).collect()
+    // We need to sort the entries to pivot them properly
+    val sorted = counts.sortBy(r => (r.get(0).toString, r.get(1).toString))
+    val first = sorted.head.get(0)
+    // get the distinct values of column 2, so that we can make them the column names
+    val distinctCol2 = sorted.takeWhile(r => r.get(0) == first)
```

Comment:
Are we assuming that the first row will contain all distinct values from col2? It is possible that the first row only has few nonzeros. If this is the case, please add a test for it.